ml-agents-envs/mlagents/envs/subprocess_env_manager.py

```diff
@@ -104,59 +109,63 @@ def __init__(
     ):
         super().__init__()
         self.env_workers: List[UnityEnvWorker] = []
+        self.step_queue: Queue = Queue()
         for worker_idx in range(n_env):
-            self.env_workers.append(self.create_worker(worker_idx, env_factory))
-
-    def get_last_steps(self):
-        return [ew.previous_step for ew in self.env_workers]
+            self.env_workers.append(
+                self.create_worker(worker_idx, self.step_queue, env_factory)
+            )
 
     @staticmethod
     def create_worker(
-        worker_id: int, env_factory: Callable[[int], BaseUnityEnvironment]
+        worker_id: int,
+        step_queue: Queue,
+        env_factory: Callable[[int], BaseUnityEnvironment],
     ) -> UnityEnvWorker:
         parent_conn, child_conn = Pipe()
 
         # Need to use cloudpickle for the env factory function since function objects aren't picklable
         # on Windows as of Python 3.6.
         pickled_env_factory = cloudpickle.dumps(env_factory)
         child_process = Process(
-            target=worker, args=(child_conn, pickled_env_factory, worker_id)
+            target=worker, args=(child_conn, step_queue, pickled_env_factory, worker_id)
         )
         child_process.start()
         return UnityEnvWorker(child_process, worker_id, parent_conn)
 
-    def step(self) -> List[StepInfo]:
+    def _queue_steps(self) -> None:
         for env_worker in self.env_workers:
-            all_action_info = self._take_step(env_worker.previous_step)
-            env_worker.previous_all_action_info = all_action_info
-            env_worker.send("step", all_action_info)
-
-        with hierarchical_timer("recv"):
-            step_brain_infos: List[AllBrainInfo] = [
-                self.env_workers[i].recv().payload for i in range(len(self.env_workers))
-            ]
-        steps = []
-        for i in range(len(step_brain_infos)):
-            env_worker = self.env_workers[i]
-            step_info = StepInfo(
-                env_worker.previous_step.current_all_brain_info,
-                step_brain_infos[i],
-                env_worker.previous_all_action_info,
-            )
-            env_worker.previous_step = step_info
-            steps.append(step_info)
-        return steps
+            if not env_worker.waiting:
+                env_action_info = self._take_step(env_worker.previous_step)
+                env_worker.previous_all_action_info = env_action_info
+                env_worker.send("step", env_action_info)
+                env_worker.waiting = True
+
+    def step(self) -> List[StepInfo]:
```

Comment:
Is a list the best way to return this? Previously the steps were 1:1 with the workers (I think), but now a worker might contribute 0 or more than 1 steps. I would think Dict[str, List[StepInfo]] where the dict key is the worker ID.

Reply:
@chriselion Could you elaborate on why you think it would be valuable to return a dictionary rather than a list?  I don't see any particular value in splitting the steps out by worker at this time since we just iterate over them elsewhere in the codebase.

It would mean changing the interface as well, which means this change wouldn't be only in the implementation of an EnvManager but bleed into TrainerController.

Reply:
OK, maybe it's not necessary.